### PR TITLE
Document public API

### LINKED ISSUE #1555: Document API struct fields
## SURE-5202

We consider several resources as public. Their API struct fields should be documented for end users.

* [GitRepo](https://fleet.rancher.io/ref-crds#gitrepo)
* [GitRepoRestriction](https://fleet.rancher.io/ref-crds#gitreporestriction)
* [Bundle](https://fleet.rancher.io/ref-crds#bundle)
* [BundleNamespaceMapping](https://fleet.rancher.io/ref-crds#bundlenamespacemapping)
* [Cluster](https://fleet.rancher.io/ref-crds#cluster)
* [ClusterGroup](https://fleet.rancher.io/ref-crds#clustergroup)
* [ClusterRegistrationToken](https://fleet.rancher.io/ref-crds#clusterregistrationtoken)
* [ImageScan](https://fleet.rancher.io/ref-crds#imagescan)
* [fleet.yaml](https://fleet.rancher.io/ref-fleet-yaml) [struct](https://github.com/rancher/fleet/blob/master/pkg/bundlereader/read.go#L133-L140)

Example: https://github.com/rancher/fleet/blob/master/pkg/apis/fleet.cattle.io/v1alpha1/bundle.go#L68-L70

We document which CRDs we consider "internal" (Bundledeployment, ClusterRegistration, Content).

### PR DESCRIPTION
**Pull request by manno**
Fix #1555

Questions for the future:
- [ ] DesiredReadyClusters for GitRepo, not accurate for new DoNotDeploy mechanic? Yes, `Bundle.Status.ResourceKey` is just a preview from helm template, `BundleDeployment.Status.Resources` is the list of actual resources per target cluster.
- [x] CorrectDrift copied into spec from options? Yes, fleet apply args vs. bundle options.
- [x] MaxNew = 50, does that work if a partition is larger? 
Moved to fleet at scale spike.
- [x] Summary checks interface{} for being a pointer?! Could be a struct.
- [x] BundleSummary and BundleDeploymentSummary, GitRepo and Cluster count bundle deployments not bundles? Yes, fleet counts targets.
- [x] confusing logic in [reset deployment](https://github.com/rancher/fleet/pull/1639/commits/3b3ce5e79dd02c3fcab6f7d8ff8b65d1c01a130d)

### CODE REVIEW

**Review by weyfonk — CHANGES_REQUESTED** [review]
Impressive work, thanks!
Please don't be scared by the large number of comments, which are mostly nitpicks. This is a great effort.
As we have a few structs which are not meant to be used outside Fleet, should we consider moving them to (a) different package(s)?

File: pkg/apis/fleet.cattle.io/v1alpha1/bundle.go [hunk 1]
```diff
@@ -253,32 +407,57 @@ type BundleDeploymentOptions struct {
...
+	Kind string `json:"kind,omitempty"`
+	// APIVersion is the apiVersion of the resource to match.
+	APIVersion string `json:"apiVersion,omitempty"`
+	// Namespace is the namespace of the resource to match.
+	Namespace string `json:"namespace,omitempty"`
+	// Name is the name of the resource to match.
+	Name string `json:"name,omitempty"`
+	// Operations remove a JSON path from the resource.
+	Operations []Operation `json:"operations,omitempty"`
+	// JSONPointers is normally empty.
```
[the remark this answers is not in the document]
Comment (manno) [reply pkg/apis/fleet.cattle.io/v1alpha1/bundle.go:426]: Ah, found this:
> // JSONPointers ignore diffs at a certain JSON path

```
    jsonPointers:
    - "/webhooks"
```

File: pkg/apis/fleet.cattle.io/v1alpha1/fleetyaml.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+// fleet.yaml is automatically turned into bundle.
+type FleetYAML struct {
+	// Name of the bundle which will be created.
+	Name string `json:"name,omitempty"`
+	// Labels are copied to the bundle and can be used in a
+	// dependsOn.selector.
+	Labels map[string]string `json:"labels,omitempty"`
+	BundleSpec
+	// TargetCustomizations are used to determine how resources should be
+	// modified per target. Targets are evaluated in order and the first
```
[the remark this answers is not in the document]
Comment (manno) [reply pkg/apis/fleet.cattle.io/v1alpha1/fleetyaml.go:14]: I thought it's a bundledeployment: `bundle+cluster=bundledeployment=target`. But you're right, the `target.go` file contains the cluster struct. Hmm.
 
Anyhow, I think together with the "first one to match a cluster is used for that cluster."  it's understandable.

File: pkg/apis/fleet.cattle.io/v1alpha1/git.go [hunk 3]
```diff
@@ -59,119 +62,188 @@ type GitRepoSpec struct {
...
+	Kind string `json:"kind,omitempty"`
+	// Type is the type of the resource, e.g. "apiextensions.k8s.io.customresourcedefinition" or "configmap".
+	Type string `json:"type,omitempty"`
+	// ID is the name of the resource, e.g. "namespace1/my-config" or "backingimagemanagers.storage.io".
+	ID string `json:"id,omitempty"`
+	// Namespace of the resource.
+	Namespace string `json:"namespace,omitempty"`
+	// Name of the resource.
+	Name string `json:"name,omitempty"`
+	// IncompleteState is true if a bundle summary has 10 or more non-ready
```
[the remark this answers is not in the document]
Comment (manno) [reply pkg/apis/fleet.cattle.io/v1alpha1/git.go:216]: Yes and not configurable. My other favorite is status.MaxNew=50 ... https://github.com/rancher/fleet/blob/document-public-crd-api/internal/cmd/controller/controllers/bundle/controller.go#L32

File: pkg/apis/fleet.cattle.io/v1alpha1/target.go [hunk 4]
```diff
@@ -41,28 +48,50 @@ type ClusterGroup struct {
...
+	// number of clusters that are ready vs. the number of clusters desired
+	// to be ready.
 	ReadyClusters string `json:"readyClusters,omitempty"`
-	ReadyBundles  string `json:"readyBundles,omitempty"`
-	State         string `json:"state,omitempty"`
+	// ReadyBundles is a string in the form "%d/%d", that described the
+	// number of bundles that are ready vs. the number of bundles desired
+	// to be ready.
+	ReadyBundles string `json:"readyBundles,omitempty"`
+	// State is a summary state for the cluster group, calculated over the non-ready resources.
```
[the remark this answers is not in the document]
Comment (manno) [reply pkg/apis/fleet.cattle.io/v1alpha1/target.go:84]: Changing this to "showing"?

File: pkg/apis/fleet.cattle.io/v1alpha1/target.go [hunk 5]
```diff
@@ -32,6 +38,7 @@ var (
 // +genclient
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
+// ClusterGroup is stored selector to target a group of clusters.
```
[the remark this answers is not in the document]
Comment (manno) [reply pkg/apis/fleet.cattle.io/v1alpha1/target.go]: You could just use the selector in `targets:`, cluster groups have little functionality besides another aggregated status struct and that you don't have to repeat your target selector in different git repos.

File: pkg/apis/fleet.cattle.io/v1alpha1/target.go [hunk 6]
```diff
@@ -116,51 +145,99 @@ type ClusterStatus struct {
...
+	// AgentDeployedGeneration is the generation of the agent that is currently deployed.
 	AgentDeployedGeneration *int64 `json:"agentDeployedGeneration,omitempty"`
-	AgentMigrated           bool   `json:"agentMigrated,omitempty"`
-	AgentNamespaceMigrated  bool   `json:"agentNamespaceMigrated,omitempty"`
-	CattleNamespaceMigrated bool   `json:"cattleNamespaceMigrated,omitempty"`
-
-	AgentAffinityHash    string `json:"agentAffinityHash,omitempty"`
-	AgentResourcesHash   string `json:"agentResourcesHash,omitempty"`
+	// AgentMigrated is always set to true after importing a cluster. If
+	// false, it will trigger a migration. Old Fleet agents, don't have
```
[the remark this answers is not in the document]
Comment (manno) [reply pkg/apis/fleet.cattle.io/v1alpha1/target.go]: Yes, it's more technical than user docs, but I find them confusing, so it's better to document something.

[the remark this answers is not in the document]
Comment (weyfonk) [reply pkg/apis/fleet.cattle.io/v1alpha1/target.go]: Fair enough.

File: pkg/apis/fleet.cattle.io/v1alpha1/bundle.go [hunk 7]
```diff
@@ -118,85 +178,179 @@ type BundleTargetRestriction struct {
...
+	// NonReadyStatus lists the state for each non-ready resource.
 	NonReadyStatus []NonReadyStatus `json:"nonReadyStatus,omitempty"`
 }
 
 var (
-	BundleConditionReady               = "Ready"
-	BundleDeploymentConditionReady     = "Ready"
+	// BundleConditionReady is unused
+	BundleConditionReady = "Ready"
+	// BundleDeploymentConditionReady will result in ErrApplied if false.
```
[the remark this answers is not in the document]
Comment (weyfonk) [reply pkg/apis/fleet.cattle.io/v1alpha1/bundle.go]: Oh I see, that makes much more sense than my interpretation. Thanks!

**Review by weyfonk — CHANGES_REQUESTED** [review]
Just one comment left on punctuation, sorry if my initial comment was not clear enough :)

**Comment by manno** [comment]
> Just one comment left on punctuation, sorry if my initial comment was not clear enough :)

Ah, I must have missed that comma.

**Pull request merged**